packages/documentsearch/src/providers/notebooksearchprovider.ts

```diff
@@ -239,7 +238,7 @@ export class NotebookSearchProvider implements ISearchProvider {
   /**
    * Report whether or not this provider has the ability to search on the given object
    */
-  static canSearchOn(domain: any): boolean {
+  static canSearchOn(domain: Widget): boolean {
```

Comment:
Would it makes sense to have the return value be a typeguard? I.e.:
```suggestion
  static canSearchOn(domain: Widget): domain is NotebookPanel {
```

Reply:
I think it would. Since this is my first brush with type guards, let me make sure I understand. If we make `canSearchOn` into a type guard, then we could take this code:

https://github.com/jupyterlab/jupyterlab/blob/ef7e7c2dbc91f6ea3a38759b459f74a2448b6be3/packages/documentsearch/src/providers/codemirrorsearchprovider.ts#L71-L82

and refactor it into the simpler:

```typescript
if (CodeMirrorSearchProvider.canSearchOn(searchTarget)) { 
  this._cm = searchTarget.content.editor;
} else {
  throw new Error('Cannot find Codemirror instance to search'); 
}
```

Is that right?

Reply:
I don't think you would need to (or should) change the ordering of the code, but you should be able to remove the type cast.

Reply:
Actually, you wouldn't be able to remove those type casts, unless you specified the generic argument of `CodeMirrorSearchProvider` further:
```ts
export type CMMainAreaWidget = MainAreaWidget &{ content: FileEditor } & {
  content: { editor: CodeMirrorEditor };
};
```

Reply:
@vidartf Great minds think similarly. I already had 

```
type CMMainAreaWidget = MainAreaWidget<FileEditor>;
```

in my latest.

> I don't think you would need to (or should) change the ordering of the code

You're right that all code following the `if` clause is aware of the type guard (I was previously under the impression that code would have to be tucked into the same `if-else` clause in order to take advantage). But more generally did you mean that it would be bad manners to change the existing code ordering?

Reply:
I finished re-implementing `canSearchOn` as a type guard and pushed it.